refactor: rework `core` and `util`

djsigmann opened the pull request.

djsigmann committed:
feat: move userdata to `userdata/` for portable installs

djsigmann committed:
refactor: rework `util`

Generally rework some of those functions and move related functions into their own submodules.

djsigmann committed:
feat: add migrations

Also add missing executable bits and shebangs to scripts.

djsigmann committed:
refactor(util): add zip submodule

djsigmann committed:
refactor(util): add file submodule

djsigmann committed:
refactor: rework auto update mechanism

djsigmann committed:
refactor(core): remove extraneous methods from `FolderConfig`

Also dynamically set attrs in `FolderConfig`.

Directories should be created/checked for existence when needed.
Using `Path` objects directly is usually cleaner than using complex methods.

djsigmann committed:
fix(handlers): remove unused import from `file_handlers.py`

cueki commented on core/operations/advanced_particle_merger.py:
```diff
@@ -121,6 +124,7 @@ def process_vpk_group(self, vpk_name: str, out_dir: Path) -> None:
 
             if elements_we_still_need:
                 game_file_path = folder_setup.temp_to_be_referenced_dir / particle_group
+                game_file_path.parent.mkdir(parents=True, exist_ok=True) # INFO: technically not necessary, but PCFFile does not check if `input_file` exists, let alone its parent dir
```
Easier to ask for forgiveness than permission, as is the case with most libraries.

djsigmann replied:
True. Removed the comment.

djsigmann committed:
refactor: various small refactors

Most notably, move all logic for downloading the modpack to `core`.

djsigmann committed:
feat(logging): redirect unhandled exceptions to `rich` if it is available

cueki commented on core/auto_updater.py:
```diff
@@ -1,187 +1,99 @@
-            tmp_path = Path(tmp_file.name)
+        match sys.platform:
+            case 'win32':
+                renamed_runme = folder_setup.install__dir / 'RUNME.tmp.bat'
```
typo im assuming... install_dir

djsigmann replied:
Yep.

djsigmann committed:
fix(packaging): specify correct version bounds for PyPI dependencies

djsigmann committed:
chore: bump to v2.0.0

djsigmann committed:
fix(workflow): make release workflow more consistent

- bump github actions' versions
- make release workflows non-concurrent
- standardize release asset archive names
